### PR TITLE
influxdb_retention_policy - add state argument to module spec (#2383)

### PR DESCRIPTION
**Pull request by xabinapal**
##### SUMMARY

Fixes: #2383

Add `state` parameter in module argument spec to support deletion of existing retention policies.

Also, simplify `duration` parsing logic as suggested on https://github.com/ansible-collections/community.general/pull/2284#discussion_r618210902.

~~A bug has also been discovered and fixed while developing the changes. It was introduced in my last PR (#2284) and it caused module failure when `duration` option was set to `INF` (a valid value).~~ (moved to #2396)

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

influxdb_retention_policy

##### ADDITIONAL INFORMATION

```yaml
- name: create retention policy
  community.general.influxdb_retention_policy:
      database_name: test
      policy_name: test
      duration: 1h
      replication: 1
      state: present

- name: drop retention policy
  community.general.influxdb_retention_policy:
      database_name: test
      policy_name: test
      state: absent
```

### CODE REVIEW

File: changelogs/fragments/2383-influxdb_retention_policy-add-state-option.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+minor_changes:
+  - influxdb_retention_policy - add ``state`` parameter with allowed values
+    ``present`` and ``absent`` to support deletion of existing retention policies
+    (https://github.com/ansible-collections/community.general/issues/2383).
+  - influxdb_retention_policy - simplify duration logic parsing as suggested in
+    https://github.com/ansible-collections/community.general/pull/2284#discussion_r618210902
+  - influxdb_retention_policy - fix bug where ``INF`` duration values failed parsing,
+    introduced in https://github.com/ansible-collections/community.general/pull/2284
```
Comment (felixfontein) [inline changelogs/fragments/2383-influxdb_retention_policy-add-state-option.yml:8]:
```suggestion
  - influxdb_retention_policy - simplify duration logic parsing
    (https://github.com/ansible-collections/community.general/pull/2385).
bugfixes:
  - influxdb_retention_policy - fix bug where ``INF`` duration values failed parsing
    (https://github.com/ansible-collections/community.general/pull/2385).
```
I would suggest to split out the bugfix into a separate PR so that this can be backported and also be used in a 3.0.1 release. If there's a feature in the PR, it can only be used in 3.1.0, but not be backported.

File: plugins/modules/database/influxdb/influxdb_retention_policy.py [hunk 2]
```diff
@@ -29,17 +29,23 @@
             - Name of the retention policy.
         required: true
         type: str
+    state:
+        description:
+            - State of the retention policy.
+        choices: [ absent, present ]
+        default: present
+        type: str
```
Comment (felixfontein) [inline plugins/modules/database/influxdb/influxdb_retention_policy.py:37]:
```suggestion
        type: str
        version_added: 3.1.0
```

File: plugins/modules/database/influxdb/influxdb_retention_policy.py [hunk 3]
```diff
@@ -29,17 +29,23 @@
...
     duration:
         description:
             - Determines how long InfluxDB should keep the data. If specified, it
               should be C(INF) or at least one hour. If not specified, C(INF) is
               assumed. Supports complex duration expressions with multiple units.
-        required: true
+            - Required only if C(state) is set to C(present).
```
Comment (felixfontein) [inline plugins/modules/database/influxdb/influxdb_retention_policy.py:43]:
```suggestion
            - Required only if I(state) is set to C(present).
```
See https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation

File: changelogs/fragments/2383-influxdb_retention_policy-add-state-option.yml (hunk 1 above)
Comment (xabinapal) [reply changelogs/fragments/2383-influxdb_retention_policy-add-state-option.yml:8]: Done, a new PR has been issued, #2396

**Comment by xabinapal** [comment]
Force-pushed to apply https://github.com/ansible-collections/community.general/pull/2385#discussion_r624133735 and https://github.com/ansible-collections/community.general/pull/2385#discussion_r624134167 and to remove bugfix commit as suggested in https://github.com/ansible-collections/community.general/pull/2385#discussion_r624133499.

Thx @felixfontein for all the suggestions 😅

File: plugins/modules/database/influxdb/influxdb_retention_policy.py [hunk 4]
```diff
@@ -208,15 +221,15 @@ def create_retention_policy(module, client):
         module.fail_json(msg="Failed to parse value of duration")
 
     influxdb_duration_format = parse_duration_literal(duration)
-    if influxdb_duration_format != 0 and influxdb_duration_format < 3600000000000:
+    if influxdb_duration_format != 0 and influxdb_duration_format < 1 * DURATION_UNIT_NANOSECS['h']:
```
Comment (russoz) [inline plugins/modules/database/influxdb/influxdb_retention_policy.py:224]: hi @xabinapal thanks for implementing those suggestions! :-) In this particular case, it *might* be slightly easier to read if we define a variable at the top of the function, like: `one_hour = 1 * DURATION_UNIT_NANOSECS['h']`, then the tests would become:
```suggestion
    if influxdb_duration_format != 0 and influxdb_duration_format < one_hour:
```

File: plugins/modules/database/influxdb/influxdb_retention_policy.py [hunk 5]
```diff
@@ -148,28 +180,9 @@ def parse_duration_literal(value, extended=False):
...
-                duration += duration_val * 1000 * 1000 * 1000 * 60
-            elif duration_literal[1] == 'h':
-                duration += duration_val * 1000 * 1000 * 1000 * 60 * 60
-            elif duration_literal[1] == 'd':
-                duration += duration_val * 1000 * 1000 * 1000 * 60 * 60 * 24
-            elif duration_literal[1] == 'w':
-                duration += duration_val * 1000 * 1000 * 1000 * 60 * 60 * 24 * 7
+        filtered_literal = list(filter(None, duration_literal))
+        duration_val = float(filtered_literal[0])
+        duration += duration_val * DURATION_UNIT_NANOSECS[filtered_literal[1]]
```
Comment (russoz) [inline plugins/modules/database/influxdb/influxdb_retention_policy.py:185]: This seems to be quite a different logic than what you had before. Could you please explain the change? I was expcting to see something like:
```suggestion
        if extended and duration_literal[3] == 's':
            duration_val = float(duration_literal[2])
        else:
            duration_val = int(duration_literal[0])
        duration += duration_val * DURATION_UNIT_NANOSECS[filtered_literal[1]]
```

**Comment by felixfontein** [comment]
Now there's a conflict (probably because 2396 has been merged), can you rebase + force push again?

File: plugins/modules/database/influxdb/influxdb_retention_policy.py (hunk 4 above)
Comment (xabinapal) [reply plugins/modules/database/influxdb/influxdb_retention_policy.py:224]: I'm now thinking that it would make sense to define that constant at the top of the file, as is used both in `create_retention_policy` and `alter_retention_policy` functions.

File: plugins/modules/database/influxdb/influxdb_retention_policy.py (hunk 5 above)
Comment (xabinapal) [reply plugins/modules/database/influxdb/influxdb_retention_policy.py:185]: Just tried to simplify code to avoid the `if` statement.

Due to how the regex is parsed, either `('1', 'm', '', '')` or `('', '', '1', 's')` tuples are returned. With `filtered_literal = list(filter(None, duration_literal))` empty strings are removed, so final tuple is always `('1', 's')`. This time I have tested this both in Python 2 and Python 3 😅 

If you think this is not clear enough, I can revert to the original `if` as you suggest.

File: changelogs/fragments/2383-influxdb_retention_policy-add-state-option.yml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+minor_changes:
+  - influxdb_retention_policy - add ``state`` parameter with allowed values
+    ``present`` and ``absent`` to support deletion of existing retention policies
+    (https://github.com/ansible-collections/community.general/issues/2383).
+  - influxdb_retention_policy - simplify duration logic parsing as suggested in
+    https://github.com/ansible-collections/community.general/pull/2284#discussion_r618210902
```
Comment (felixfontein) [inline changelogs/fragments/2383-influxdb_retention_policy-add-state-option.yml:6]:
```suggestion
  - influxdb_retention_policy - simplify duration logic parsing
    (https://github.com/ansible-collections/community.general/pull/2385).
```

File: plugins/modules/database/influxdb/influxdb_retention_policy.py [hunk 7]
```diff
@@ -272,30 +287,58 @@ def alter_retention_policy(module, client, retention_policy):
...
         default=dict(default=False, type='bool'),
         shard_group_duration=dict(required=False, type='str'),
     )
     module = AnsibleModule(
         argument_spec=argument_spec,
-        supports_check_mode=True
+        supports_check_mode=True,
+        required_if=(
+            ('state', 'present', ['duration', 'replication']),
+        )
```
Comment (felixfontein) [inline plugins/modules/database/influxdb/influxdb_retention_policy.py:321]:
```suggestion
        ),
```
If you add this comma, the next time someone adds another kwarg this line does not need to be touched :)

File: plugins/modules/database/influxdb/influxdb_retention_policy.py [hunk 8]
```diff
@@ -272,30 +287,58 @@ def alter_retention_policy(module, client, retention_policy):
...
 def main():
     argument_spec = InfluxDb.influxdb_argument_spec()
     argument_spec.update(
+        state=dict(default='present', type='str', choices=['present', 'absent']),
         database_name=dict(required=True, type='str'),
         policy_name=dict(required=True, type='str'),
-        duration=dict(required=True, type='str'),
-        replication=dict(required=True, type='int'),
+        duration=dict(required=False, type='str'),
+        replication=dict(required=False, type='int'),
```
Comment (felixfontein) [inline plugins/modules/database/influxdb/influxdb_retention_policy.py:312]:
```suggestion
        duration=dict(type='str'),
        replication=dict(type='int'),
```

File: plugins/modules/database/influxdb/influxdb_retention_policy.py [hunk 9]
```diff
@@ -272,30 +287,58 @@ def alter_retention_policy(module, client, retention_policy):
...
+
+    changed = False
+
+    if not module.check_mode:
+        try:
+            client.drop_retention_policy(policy_name, database_name)
+        except exceptions.InfluxDBClientError as e:
+            module.fail_json(msg=e.content)
+        changed = True
+    module.exit_json(changed=changed)
```
Comment (felixfontein) [inline plugins/modules/database/influxdb/influxdb_retention_policy.py:302]: In check mode, you must also return `changed=True`:
```suggestion
    module.exit_json(changed=True)
```

File: plugins/modules/database/influxdb/influxdb_retention_policy.py (hunk 8 above)
Comment (xabinapal) [reply plugins/modules/database/influxdb/influxdb_retention_policy.py:312]: There is a third argument, `shard_group_duration`, not modified in this PR, that also has `required=False`. I added it in both `duration` and `replication` to "standardize" format. Should I remove it from that argument too?

Comment (felixfontein) [reply plugins/modules/database/influxdb/influxdb_retention_policy.py:309]: Yes, please modify that one as well. Thanks!

File: plugins/modules/database/influxdb/influxdb_retention_policy.py (hunk 5 above)
Comment (russoz) [reply plugins/modules/database/influxdb/influxdb_retention_policy.py:187]: Oh OK, I hadn't examined it deeply enough. I think it is good as it is.

**Review by russoz — APPROVED** [review]
Looking good to me, thanks for implementing!

File: plugins/modules/database/influxdb/influxdb_retention_policy.py [hunk 10]
```diff
@@ -134,6 +153,21 @@
 DURATION_REGEX = re.compile(r'(\d+)(ns|u|µ|ms|s|m|h|d|w)')
 EXTENDED_DURATION_REGEX = re.compile(r'(?:(\d+)(ns|u|µ|ms|m|h|d|w)|(\d+(?:\.\d+)?)(s))')
 
+DURATION_UNIT_NANOSECS = {
+    'ns': 1,
+    'u': 1000,
+    'µ': 1000,
```
Comment (aminvakil) [inline plugins/modules/database/influxdb/influxdb_retention_policy.py:159]: I'm not sure about this, could adding `µ` possibly cause some issues later on, formatting-wise?

**Review by aminvakil — APPROVED** [review]
Thanks for adding this!
I just put a comment which I don't think would be a blocker for merging this.

File: plugins/modules/database/influxdb/influxdb_retention_policy.py (hunk 10 above)
Comment (xabinapal) [reply plugins/modules/database/influxdb/influxdb_retention_policy.py:159]: On #2383 `µ` was already included, this is only a rework to simplify that part of code.

Just out of curiosity, what kind of issues could it cause? I think it's an standard character in all encodings (I've checked extended ASCII, UTF-8, Windows-1252) but it's true that no one in their right mind would use `µ` instead of `u`. I just included it to be fully compatible with InfluxDB spec.

Comment (aminvakil) [reply plugins/modules/database/influxdb/influxdb_retention_policy.py:159]: OK then, thanks for checking!

**Comment by felixfontein** [comment]
If nobody complains, I'll merge this soon.

File: plugins/modules/database/influxdb/influxdb_retention_policy.py [hunk 11]
```diff
@@ -110,6 +121,14 @@
       ssl: no
       validate_certs: no
       shard_group_duration: 1d10h30m
+      state: present
+
+- name: Drop retention policy
+  community.general.influxdb_retention_policy:
+      hostname: "{{influxdb_ip_address}}"
```
[the remark this answers is not in the document]
Comment (xabinapal) [reply plugins/modules/database/influxdb/influxdb_retention_policy.py:128]: I suppose you meant:
```suggestion
      hostname: "{{ influxdb_ip_address }}"
```

In fact I agree that this increases readability, in my job I tend to do it that way.

I don't mind changing it, I just followed the style used in the rest of the examples. If this style is preferred, should I change all the examples?

[the remark this answers is not in the document]
Comment (felixfontein) [reply plugins/modules/database/influxdb/influxdb_retention_policy.py:128]: Yes, please change all of them :)

[the remark this answers is not in the document]
Comment (aminvakil) [reply plugins/modules/database/influxdb/influxdb_retention_policy.py:128]: > I suppose you meant:
> 
> In fact I agree that this increases readability, in my job I tend to do it that way.

Ah, sure, I changed my suggestions as well.
> I don't mind changing it, I just followed the style used in the rest of the examples. If this style is preferred, should I change all the examples?

If you don't mind, yes please.

[the remark this answers is not in the document]
Comment (xabinapal) [reply plugins/modules/database/influxdb/influxdb_retention_policy.py]: There you go, it's done.

I've checked also #2368 to change those module examples too but they already had whitespaces, so nothing to do 😄

**Comment by felixfontein** [comment]
@xabinapal thanks for improving this!
@russoz @aminvakil thanks for reviewing!
